cdap-proto/src/main/java/co/cask/cdap/proto/security/Principal.java

```diff
@@ -38,8 +38,10 @@
 
   private final String name;
   private final PrincipalType type;
-  private final int hashCode;
   private final String kerberosPrincipal;
+  // This needs to be transient because we don't want this to be populated during deserialization since that
+  // value will not be provided by the client.
+  private transient Integer hashCode;
```

Comment:
It can, but doesn't need to. This is because the hash code is computed from immutable fields. 
Yes, it's possible to have more then one computations across concurrent threads. If the object.hashCode get used many times, the cost of accessing a non-volatile field may out-weight the cost of small number of recomputation, as any new thread started after the current threads will see the value.